File: exercises/05-pubsub-go/ex01_publish_subscribe.go
// Exercise 01 — Pub/Sub: Publish and Subscribe (Go)
// ---------------------------------------------------
// Goal:
//   - Publish messages to a Pub/Sub topic using Go
//   - Pull messages with a subscriber callback
//   - Handle concurrent message processing safely
//   - Use context cancellation for clean shutdown
//
// Run:
//   go mod init pubsub-go-exercises
//   go get cloud.google.com/go/pubsub
//   export GOOGLE_CLOUD_PROJECT=your-project-id
//
//   gcloud pubsub topics create events-topic
//   gcloud pubsub subscriptions create events-sub --topic=events-topic
//
//   go run ex01_publish_subscribe.go

package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"os"
	"sync"
	"sync/atomic"
	"time"

	"cloud.google.com/go/pubsub"
)

// ---------------------------------------------------------------------------
// Config
// ---------------------------------------------------------------------------
var (
	projectID      = os.Getenv("GOOGLE_CLOUD_PROJECT")
	topicID        = "events-topic"
	subscriptionID = "events-sub"
)

// ---------------------------------------------------------------------------
// Event — payload struct for messages
// ---------------------------------------------------------------------------
type Event struct {
	EventID   string  `json:"event_id"`
	EventType string  `json:"event_type"`
	Country   string  `json:"country"`
	UserID    string  `json:"user_id"`
	Amount    float64 `json:"amount"`
	Timestamp int64   `json:"timestamp"`
}

func main() {
	if projectID == "" {
		log.Fatal("GOOGLE_CLOUD_PROJECT not set")
	}

	ctx := context.Background()
	client, err := pubsub.NewClient(ctx, projectID)
	if err != nil {
		log.Fatalf("pubsub.NewClient: %v", err)
	}
	defer client.Close()

	// Run exercises
	setupTopicAndSubscription(ctx, client)
	publishEvents(ctx, client, 30)

	time.Sleep(2 * time.Second) // Allow messages to be available

	pullMessagesConcurrently(ctx, client, 15*time.Second)
}

// ---------------------------------------------------------------------------
// EXERCISE 1a: Create topic and subscription
// ---------------------------------------------------------------------------
func setupTopicAndSubscription(ctx context.Context, client *pubsub.Client) {
	// TODO: Get or create the topic
	// client.CreateTopic(ctx, topicID) returns (*pubsub.Topic, error)
	topic, err := client.CreateTopic(ctx, topicID)
	if err != nil {
		// Topic likely already exists
		topic = client.Topic(topicID)
		fmt.Printf("Topic %s: using existing\n", topicID)
	} else {
		fmt.Printf("Topic %s: created\n", topic.String())
	}
	_ = topic

	// TODO: Get or create the subscription
	sub, err := client.CreateSubscription(ctx, subscriptionID, pubsub.SubscriptionConfig{
		Topic:            client.Topic(topicID),
		AckDeadline:      30 * time.Second,
		RetentionDuration: 24 * time.Hour,
	})
	if err != nil {
		sub = client.Subscription(subscriptionID)
		fmt.Printf("Subscription %s: using existing\n", subscriptionID)
	} else {
		fmt.Printf("Subscription %s: created\n", sub.String())
	}
}

// ---------------------------------------------------------------------------
// EXERCISE 1b: Publish events concurrently (using goroutines + futures)
// ---------------------------------------------------------------------------
func publishEvents(ctx context.Context, client *pubsub.Client, count int) {
	topic := client.Topic(topicID)
	// Set batching settings for throughput
	topic.PublishSettings.ByteThreshold  = 5000
	topic.PublishSettings.CountThreshold = 10
	topic.PublishSettings.DelayThreshold = 100 * time.Millisecond
	defer topic.Stop() // Flush remaining messages on exit

	actions   := []string{"purchase", "click", "view", "add_to_cart", "refund"}
	countries := []string{"US", "GB", "DE", "JP", "BR"}

	var wg sync.WaitGroup
	var published int64

	for i := 0; i < count; i++ {
		event := Event{
			EventID:   fmt.Sprintf("evt-%05d", i),
			EventType: actions[rand.Intn(len(actions))],
			Country:   countries[rand.Intn(len(countries))],
			UserID:    fmt.Sprintf("user-%04d", rand.Intn(50)+1),
			Amount:    float64(rand.Intn(20000)) / 100.0,
			Timestamp: time.Now().UnixMilli(),
		}

		data, err := json.Marshal(event)
		if err != nil {
			log.Printf("json.Marshal: %v", err)
			continue
		}

		// TODO: Publish the message with attributes
		// topic.Publish(ctx, &pubsub.Message{...}) returns a *pubsub.PublishResult
		result := topic.Publish(ctx, &pubsub.Message{
			Data: data,
			Attributes: map[string]string{
				"event_type": event.EventType,
				"country":    event.Country,
			},
		})

		wg.Add(1)
		go func(r *pubsub.PublishResult, id string) {
			defer wg.Done()
			// TODO: Block until the publish completes and get the server-assigned message ID
			_, err := r.Get(ctx)
			if err != nil {
				log.Printf("Publish failed for %s: %v", id, err)
				return
			}
			atomic.AddInt64(&published, 1)
		}(result, event.EventID)
	}

	wg.Wait()
	fmt.Printf("Published %d/%d events to %s\n", published, count, topicID)
}

// ---------------------------------------------------------------------------
// EXERCISE 1c: Pull messages with a concurrent subscriber callback
// ---------------------------------------------------------------------------
func pullMessagesConcurrently(ctx context.Context, client *pubsub.Client, duration time.Duration) {
	sub := client.Subscription(subscriptionID)

	// Configure concurrency — how many messages to process in parallel
	// TODO: Set sub.ReceiveSettings to control concurrency
	sub.ReceiveSettings.MaxOutstandingMessages = 50
	sub.ReceiveSettings.NumGoroutines          = 4  // parallel goroutines

	var (
		received int64
		mu       sync.Mutex
		counts   = make(map[string]int)
	)

	// Timeout context — stop pulling after `duration`
	receiveCtx, cancel := context.WithTimeout(ctx, duration)
	defer cancel()

	fmt.Printf("\n=== Pulling messages for %v ===\n", duration)

	// TODO: Call sub.Receive(ctx, callback) — blocks until ctx is cancelled
	// The callback is called concurrently for each received message
	err := sub.Receive(receiveCtx, func(ctx context.Context, msg *pubsub.Message) {
		var event Event
		if err := json.Unmarshal(msg.Data, &event); err != nil {
			log.Printf("json.Unmarshal error: %v — nacking", err)
			// TODO: Nack the message to trigger redelivery
			msg.Nack()
			return
		}

		atomic.AddInt64(&received, 1)
		mu.Lock()
		counts[event.EventType]++
		mu.Unlock()

		// TODO: Acknowledge the message after successful processing
		msg.Ack()

		// Simulate processing time
		time.Sleep(10 * time.Millisecond)
	})

	if err != nil && err != context.DeadlineExceeded {
		log.Printf("Receive error: %v", err)
	}

	fmt.Printf("\nPull complete. Total received: %d\n", received)
	fmt.Println("Events by type:")
	for eventType, count := range counts {
		fmt.Printf("  %-15s %d\n", eventType, count)
	}
}

// ---------------------------------------------------------------------------
// CHALLENGES
// ---------------------------------------------------------------------------
// 1. Add a PublishResult error handler that retries failed publishes
//    up to 3 times with exponential backoff.
//
// 2. Implement a metrics counter: track messages/second received.
//    Sample the atomic counter every second in a goroutine.
//
// 3. Create a filtered subscription (server-side filtering):
//    pubsub.SubscriptionConfig{
//        Filter: `attributes.event_type = "purchase"`,
//    }
//    Compare how many messages you receive vs without the filter.
//
// 4. Research: what is Pub/Sub "exactly-once delivery" mode?
//    How is it different from the default "at-least-once"?
//    When would you enable it and what's the trade-off?

File: exercises/04-pubsub-python/ex01_publish_subscribe.py
"""
Exercise 01 — Pub/Sub: Publish and Subscribe (Python)
------------------------------------------------------
Goal:
    - Publish messages to a Pub/Sub topic
    - Pull messages from a subscription
    - Use streaming pull (push-style async consumer)
    - Handle message acknowledgement and error cases

Interview relevance:
    "What is Pub/Sub and how does it fit in a GCP streaming pipeline?"
    Answer: Pub/Sub is the event ingestion layer — like Kafka on GCP.
    Producers publish → Pub/Sub buffers → Consumers (Dataflow, Dataproc, Cloud Functions) pull

Setup:
    pip install google-cloud-pubsub
    export GOOGLE_CLOUD_PROJECT=your-project-id

    Create topic and subscription:
    gcloud pubsub topics create events-topic
    gcloud pubsub subscriptions create events-sub --topic=events-topic
"""

import os
import json
import time
import uuid
import random
import threading
from concurrent.futures import TimeoutError
from google.cloud import pubsub_v1

PROJECT_ID    = os.environ.get("GOOGLE_CLOUD_PROJECT", "your-project-id")
TOPIC_ID      = "events-topic"
SUBSCRIPTION_ID = "events-sub"


# ---------------------------------------------------------------------------
# EXERCISE 1a: Create topic and subscription programmatically
# ---------------------------------------------------------------------------
def setup_topic_and_subscription() -> None:
    """
    Create a Pub/Sub topic and pull subscription via the Admin API.
    In production you'd do this with Terraform (see exercises/07-terraform-gcp/).
    """
    publisher   = pubsub_v1.PublisherClient()
    subscriber  = pubsub_v1.SubscriberClient()

    topic_path = publisher.topic_path(PROJECT_ID, TOPIC_ID)
    sub_path   = subscriber.subscription_path(PROJECT_ID, SUBSCRIPTION_ID)

    # TODO: Create topic — ignore AlreadyExists error
    try:
        publisher.create_topic(request={"name": topic_path})
        print(f"Topic created: {topic_path}")
    except Exception as e:
        print(f"Topic exists or error: {e}")

    # TODO: Create subscription pointing to the topic
    try:
        subscriber.create_subscription(request={
            "name": sub_path,
            "topic": topic_path,
            "ack_deadline_seconds": 30,
        })
        print(f"Subscription created: {sub_path}")
    except Exception as e:
        print(f"Subscription exists or error: {e}")


# ---------------------------------------------------------------------------
# EXERCISE 1b: Publish messages to a topic
# ---------------------------------------------------------------------------
def publish_events(num_events: int = 20) -> None:
    """
    Publish JSON-encoded event messages to a Pub/Sub topic.

    Key concepts:
    - Messages are bytes — encode your payload (JSON, Avro, Protobuf)
    - Attributes are key-value string metadata on each message
    - publisher.publish() is async — returns a future
    - Call future.result() to block until published (or timeout)
    """
    publisher  = pubsub_v1.PublisherClient()
    topic_path = publisher.topic_path(PROJECT_ID, TOPIC_ID)

    actions   = ["purchase", "click", "view", "add_to_cart", "refund"]
    countries = ["US", "GB", "DE", "JP", "BR"]

    futures = []
    for i in range(num_events):
        event = {
            "event_id":   str(uuid.uuid4()),
            "event_type": random.choice(actions),
            "country":    random.choice(countries),
            "user_id":    f"user_{random.randint(1, 50):04d}",
            "amount":     round(random.uniform(1.0, 200.0), 2),
            "timestamp":  time.time(),
        }

        # Encode the payload as bytes (Pub/Sub requires bytes)
        data = json.dumps(event).encode("utf-8")

        # TODO: Publish the message with attributes
        # publisher.publish(topic_path, data=data, event_type=event["event_type"])
        future = publisher.publish(
            topic_path,
            data=data,
            # Attributes are string key-value pairs — useful for filtering
            event_type=event["event_type"],
            country=event["country"],
        )
        futures.append((future, event["event_id"]))

    # Wait for all publishes to complete
    for future, event_id in futures:
        try:
            message_id = future.result(timeout=10)
            # print(f"  Published {event_id} → message_id={message_id}")
        except Exception as e:
            print(f"  Publish failed for {event_id}: {e}")

    print(f"Published {num_events} events to {topic_path}")


# ---------------------------------------------------------------------------
# EXERCISE 1c: Pull messages synchronously (simple polling)
# ---------------------------------------------------------------------------
def pull_messages_sync(max_messages: int = 10) -> None:
    """
    Synchronous pull — pull up to max_messages at once, process, acknowledge.
    Good for batch processing or scheduled consumers.

    NOT recommended for high-throughput — use streaming pull instead.
    """
    subscriber = pubsub_v1.SubscriberClient()
    sub_path   = subscriber.subscription_path(PROJECT_ID, SUBSCRIPTION_ID)

    print(f"\n=== Synchronous pull (up to {max_messages} messages) ===")

    with subscriber:
        # TODO: Call subscriber.pull(request={...}) with max_messages
        response = subscriber.pull(
            request={
                "subscription": sub_path,
                "max_messages":  max_messages,
            },
            timeout=10,
        )

        ack_ids = []
        for msg in response.received_messages:
            payload = json.loads(msg.message.data.decode("utf-8"))
            attrs   = dict(msg.message.attributes)
            print(f"  [{msg.message.message_id}] {payload['event_type']:<15} "
                  f"country={attrs.get('country', '??')}  "
                  f"amount=${payload['amount']:.2f}")
            ack_ids.append(msg.ack_id)

        # TODO: Acknowledge the messages so they are not redelivered
        if ack_ids:
            subscriber.acknowledge(request={"subscription": sub_path, "ack_ids": ack_ids})
            print(f"Acknowledged {len(ack_ids)} messages.")
        else:
            print("No messages available.")


# ---------------------------------------------------------------------------
# EXERCISE 1d: Streaming pull (async subscriber — production pattern)
# ---------------------------------------------------------------------------
def streaming_pull(timeout_seconds: int = 15) -> None:
    """
    Streaming pull — the recommended production pattern.
    The subscriber library manages a long-lived connection and delivers
    messages to a callback as they arrive.

    Key behaviour:
    - callback() is called in a thread pool for each message
    - You MUST call message.ack() or message.nack() in the callback
    - nack() = redeliver after ack_deadline expires
    """
    subscriber = pubsub_v1.SubscriberClient()
    sub_path   = subscriber.subscription_path(PROJECT_ID, SUBSCRIPTION_ID)

    received_count = {"n": 0}
    lock = threading.Lock()

    def callback(message: pubsub_v1.types.PubsubMessage) -> None:
        """
        Called for each incoming message.
        Process the message and ack (or nack on failure).
        """
        try:
            payload = json.loads(message.data.decode("utf-8"))
            with lock:
                received_count["n"] += 1
                print(f"  Received #{received_count['n']}: "
                      f"{payload['event_type']:<15} "
                      f"user={payload['user_id']}  "
                      f"amount=${payload['amount']:.2f}")

            # TODO: Acknowledge the message after successful processing
            message.ack()

        except Exception as e:
            print(f"  Processing failed: {e} — nacking")
            # TODO: Nack on failure — message will be redelivered
            message.nack()

    # Flow control: limit inflight messages to avoid memory issues
    flow_control = pubsub_v1.types.FlowControl(max_messages=100)

    print(f"\n=== Streaming pull (running for {timeout_seconds}s) ===")
    with subscriber:
        streaming_pull_future = subscriber.subscribe(
            sub_path,
            callback=callback,
            flow_control=flow_control,
        )
        try:
            streaming_pull_future.result(timeout=timeout_seconds)
        except TimeoutError:
            streaming_pull_future.cancel()
            streaming_pull_future.result()

    print(f"Streaming pull finished. Received {received_count['n']} messages.")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    print("=== Pub/Sub Python Exercise ===\n")
    setup_topic_and_subscription()

    print("\n--- Publishing events ---")
    publish_events(num_events=30)

    time.sleep(2)   # Brief wait for messages to be available

    print("\n--- Sync pull ---")
    pull_messages_sync(max_messages=10)

    print("\n--- Streaming pull ---")
    # Publish more events for the streaming pull to consume
    publish_events(num_events=20)
    streaming_pull(timeout_seconds=10)

# ---------------------------------------------------------------------------
# CHALLENGES
# ---------------------------------------------------------------------------
# 1. Add a filter subscription: create a subscription with a filter
#    that only delivers messages where event_type = "purchase".
#    How does the subscriber.create_subscription request change?
#    Hint: add "filter": 'attributes.event_type = "purchase"' to the request.
#
# 2. Implement a dead-letter topic: create a second topic "events-dlq"
#    and configure the subscription to send unacked messages there after
#    5 delivery attempts. Simulate a nack loop and watch messages move to DLQ.
#
# 3. Measure end-to-end latency: record time.time() in the published message,
#    and subtract it in the callback. What's the typical Pub/Sub latency?
#
# 4. Research: in a real pipeline, the Dataflow Beam job would replace
#    your streaming_pull() callback. Pub/Sub → Dataflow = fully managed.
#    What are the advantages over a custom streaming_pull consumer?
